tools/codemod.py

```diff
@@ -0,0 +1,242 @@
+    ret_value = unparse(ret)
+    if return_transform is None:
+        yield ret_value
+    elif return_transform == "assign":
+        yield f"{assign_to} = {ret_value}"
+    else:
+        raise TypeError(return_transform)
+
+
+def embed_extract_func_def(
+    module_name: str,
+    func_name: str,
+    /,
+    before_code: _Code | None = None,
+    after_code: _Code | None = None,
+    assign_to: str = "chart",
+    indent: int | None = None,
+):
+    parts: list[_Code] = []
+    if before_code is not None:
+        parts.append(before_code)
+    parts.append(
+        iter_func_def_unparse(
+            module_name, func_name, return_transform="assign", assign_to=assign_to
+        )
+    )
+    if after_code is not None:
+        parts.append(after_code)
+    normed = "\n".join(normalize_code(s) for s in parts)
+    checked = ruff_check_str(normed)
+    formatted = ruff_format_str(checked)
+    return textwrap.indent(formatted, " " * indent) if indent else formatted
+
+
+def extract_func_def(
+    module_name: str,
+    func_name: str,
+    *,
+    format: bool = True,
+    output: Literal["altair-plot", "code-block", "str"] = "str",
+) -> str:
+    """
+    Extract the contents of a function for use as a code block.
+
+    Parameters
+    ----------
+    module_name
+        Absolute, dotted import style.
+    func_name
+        Name of function in ``module_name``.
+    format
+        Run through ``ruff format`` before returning.
+    output
+        Optionally, return embedded in an `rst` directive.
+
+    Notes
+    -----
+    - Functions must define all imports inline, to ensure they are propagated
+    - Must end with a single return statement
+
+    Warning
+    -------
+    Requires ``python>=3.9`` for `ast.unparse`_
+
+    Examples
+    --------
+    Transform the contents of a function into a code block::
+
+        >>> extract_func_def("tests.altair_theme_test", "alt_theme_test", output="code-block") # doctest: +SKIP
+
+    .. _ast.unparse:
+        https://docs.python.org/3.9/library/ast.html#ast.unparse
+    """
+    if output not in {"altair-plot", "code-block", "str"}:
+        raise TypeError(output)
+    it = iter_func_def_unparse(module_name, func_name)
+    s = ruff_format_str(it, trailing_comma=False) if format else "\n".join(it)
+    if output == "str":
+        return s
+    else:
+        return f".. {output}::\n\n{textwrap.indent(s, ' ' * 4)}\n"
+
+
+# TODO: Move these and `tools.schemapi.utils.ruff_` into a `Ruff` class
+# - Then can support configuring an instance like the `mistune` classes
+def ruff_check_str(code: _Code, /) -> str:
+    encoded = normalize_code(code).encode()
+    cmd = [
+        "ruff",
+        "check",
+        "--fix",
+        "--ignore",
+        "E711",  # Comparison to `None`
+        "--stdin-filename",
+        "placeholder.py",
+    ]
+    capture_output = True
+    r = subprocess.run(cmd, input=encoded, check=True, capture_output=capture_output)
+    return r.stdout.decode()
```

Comment:
## TODO
Move these into a `tools.codemod.Ruff` class

- `tools.codemod.ruff_check_str`
- `tools.codemod.ruff_format_str`
- `tools.schemapi.utils.ruff_format_py`
- `tools.schemapi.utils.ruff_write_lint_format_str`

Then can support configuring an instance like the `mistune` classes in `tools.markup`

Reply:
Resolved in [refactor: Adds `tools.codemod.Ruff`](https://github.com/vega/altair/pull/3630/commits/06152280cc4e993eb12bbcc2b604af247c56ae21)